### PR TITLE
Use MultiGet in RocksDB Materializer.

### PR DESCRIPTION
**Pull request by maierlars**
### Scope & Purpose
Use MultiGet in the Materialize Executor to hopefully speed up execution time.

The current implementation is not ideal. Using `PhysicalCollection::lookup` with multiple keys is a nice API to have, but it requires allocations and reprocessing of the keys. In the final versoin the MaterializeExecutor should directly use the `RocksDBMethods` and prepare keys by itself.

TODOs:
- [ ] Memory Accounting

### CODE REVIEW

File: arangod/Aql/MaterializeExecutor.h [hunk 1]
```diff
@@ -120,6 +120,8 @@ class MaterializeRocksDBExecutor : public MaterializeExecutorBase {
 
  private:
   PhysicalCollection* _collection{};
+  std::vector<LocalDocumentId> _docIds;
+  std::vector<InputAqlItemRow> _inputRows;
```
Comment (MBkkt) [inline arangod/Aql/MaterializeExecutor.h:124]: There is no reason to have more than 32 keys in the MultiGet call.
You can look MultiGet wrapper used for search

File: arangod/Aql/OptimizerRuleBatchMaterializeDocuments.cpp [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+    }
+    if (index->canReadOwnWrites() == ReadOwnWrites::yes) {
+      LOG_RULE << "INDEX " << index->id() << " FAILED: "
+               << "index has to read its own write - not supported";
+      continue;
+    }
+
+    if (index->estimateCost().estimatedNrItems < 100) {
+      LOG_RULE << "INDEX " << index->id() << " FAILED: "
+               << "estimated number of items to small";
```
Comment (jsteemann) [inline arangod/Aql/OptimizerRuleBatchMaterializeDocuments.cpp:101]:
```suggestion
               << "estimated number of items too small";
```

File: arangod/Aql/OptimizerRulesJoin.cpp [hunk 3]
```diff
@@ -417,6 +417,12 @@ void findCandidates(IndexNode* indexNode,
         calculations.push_back(calc);
         parent = parent->getFirstParent();
         continue;
+      } else if (parent->getType() == EN::MATERIALIZE) {
+        // we can always move past materialize nodes
+        // auto mat =
+        // ExecutionNode::castTo<materialize::MaterializeNode*>(parent);
```
Comment (jsteemann) [inline arangod/Aql/OptimizerRulesJoin.cpp:423]:
```suggestion
```

File: arangod/Aql/JoinExecutor.cpp [hunk 4]
```diff
@@ -61,14 +61,14 @@ void JoinExecutorInfos::determineProjectionsForRegisters() {
 JoinExecutor::~JoinExecutor() {
   // handle decrease of memory usage of the projections builder
   if (!_projectionsBuilder.isEmpty()) {
-    resourceMonitor().decreaseMemoryUsage(_projectionsBuilder.size());
+    // resourceMonitor().decreaseMemoryUsage(_projectionsBuilder.size());
```
Comment (jsteemann) [inline arangod/Aql/JoinExecutor.cpp:64]: I suggest cleaning this (and following occurrences) up to make it easier to follow where memory management happens.

**Review by jsteemann — APPROVED** [review]
LGTM

**Pull request merged**